### PR TITLE
2022.04.03 Project 1 Scissors Paper Stone Kaiquan

### PR DESCRIPTION
**Pull request by KaiquanMah**

### CODE REVIEW

**Comment by KaiquanMah** [comment]
Main files to review for Project 1 Scissors Paper Stone (in the project1-sps folder):
-index.html
-script.js

- How many hours did you spend on this assignment? 4 hours I think.
- What part of the assignment did you spend the most time on? Refactoring the code and testing to see whether the refactored code works and links properly.
- What is one error and/or error message you received while working on this assignment, and how did you solve the error? Sometimes output didnt come out. I had to test the functions in the browser's console
- How comfortable did you feel with this assignment? (1-5): 3
- Is there anything in this code that you feel pleased about? Being able to get a refactored version of my code
- What's one aspect of your code you would like specific, elaborate feedback on? Are there ways to simplify refactoring of codes?

**Comment by KaiquanMah** [comment]
Main file to review for Project 2 Beat That (in the project2-beat-that folder):
-script.js

- How many hours did you spend on this assignment? 4 hours.
- What part of the assignment did you spend the most time on? Understanding how to go from 2 dice throws per person to the dice order and the final number based on the dice order.
- What is one error and/or error message you received while working on this assignment, and how did you solve the error? Cant remember
- How comfortable did you feel with this assignment? (1-5): 3
- Is there anything in this code that you feel pleased about? n.a.
- What's one aspect of your code you would like specific, elaborate feedback on? n.a.

File: project1-sps/script.js [hunk 1]
```diff
@@ -1,4 +1,144 @@
...
+    userWin = 0.5;
+  }
+
+  return userWin;
+};
+
+// Refactored main fn
+var main = function (userInput) {
+  // welcome a user
+  if (boolUserEnteredName === 0 && userInput != "") {
```
Comment (lee-kimmixq) [inline project1-sps/script.js:66]: What happens if user has not entered name but input is also empty?

File: project1-sps/script.js [hunk 2]
```diff
@@ -1,4 +1,144 @@
...
+    return `Welcome ${userInput}! 
+    <br><br>
+    Now enter "scissors", "paper" or "stone" in the "Input" box and click
+    "Submit" to play the game!`;
+  }
+  // play the game
+  else if (boolUserEnteredName === 1) {
+    // validate user input
+    let userWrongInput = 0;
+    if (
```
Comment (lee-kimmixq) [inline project1-sps/script.js:78]: Good input validation!

File: project1-sps/script.js [hunk 3]
```diff
@@ -1,4 +1,144 @@
...
+      userWrongInput = 1;
+    }
+
+    // Generate random int between 1 to 3 (inclusive)
+    const maxNum = 3;
+    let iRandomInt = randomIntegerGenerator(maxNum);
+
+    // Retrieve program's random choice of 'scissors/paper/stone'
+    let sComputerRandomChoice = "";
+    sComputerRandomChoice = retrieveComputerRandomChoice(iRandomInt);
```
Comment (lee-kimmixq) [inline project1-sps/script.js:93]: Since you are assigning a value to the variable immediately after defining it, you can shorten these 2 lines to `const sComputerRandomChoice = retrieveComputerRandomChoice(iRandomInt);`

File: project1-sps/script.js [hunk 4]
```diff
@@ -1,4 +1,144 @@
...
+    // console.log("userWin", userWin);
+    userWin = computeUserWin(userInput, sComputerRandomChoice, userWin);
+
+    if (userWrongInput != 1) {
+      iCountUserWins += Math.floor(userWin);
+      if (userWin === 1) {
+        iCountComputerWins += 1;
+      } else if (userWin === 0.5) {
+        iCountDraws += 1;
+      }
```
Comment (lee-kimmixq) [inline project1-sps/script.js:109]: Interesting way of tallying score!

File: project1-sps/script.js [hunk 5]
```diff
@@ -1,4 +1,144 @@
...
+      return `The computer chose ${sComputerRandomChoice}.
+      You chose ${userInput}.
+      You lost! 
+      <br><br>
+      Try harder next round. So far ${userName}, you've been winning ${iCountUserWins}/${iCountTotalRounds} rounds, with ${iCountDraws} draws.
+      <br><br>
+      Now you can type "scissors" "paper" or "stone" to play another round.`;
+    } else {
+      return "EXIT_ERROR";
+    }
```
Comment (lee-kimmixq) [inline project1-sps/script.js:142]: Very neat block of all the output messages and template literals! We can also try to make this shorter - note the similarities between the game results messages (The computer chose... You chose...). We can try not to repeat those portions, and only change the message between the line breaks. You can also refactor this into a `makeResultMessage(userWin)` helper function!

File: project1-sps/script.js [hunk 6]
```diff
@@ -1,4 +1,144 @@
...
+
+var randomIntegerGenerator = function (maxNum) {
+  let nRandomNum = Math.random() * maxNum;
+  let iRandomInt = Math.floor(nRandomNum);
+  iRandomInt += 1;
+  return iRandomInt;
+};
+
+// Helper function to retrieve computer random choice from computer's random integer
+var retrieveComputerRandomChoice = function (iRandomInt) {
```
Comment (lee-kimmixq) [inline project1-sps/script.js:17]: Nice helper function and good naming :)

File: project1-sps/script.js [hunk 7]
```diff
@@ -1,4 +1,144 @@
...
+    sComputerRandomChoice = "stone";
+  } else {
+    sComputerRandomChoice = "error";
+  }
+
+  return sComputerRandomChoice;
+};
+
+// Helper function to compute userWin (0 means loss, 0.5 means draw, 1 means won)
+var computeUserWin = function (userInput, sComputerRandomChoice, userWin) {
```
Comment (lee-kimmixq) [inline project1-sps/script.js:33]: Nice helper function and good naming 2 :)

File: project1-sps/script.js [hunk 8]
```diff
@@ -1,4 +1,144 @@
...
+    userWin = 0;
+  }
+  // If both parties choose the same object, it's a draw (assign arbitrary score of 0.5)
+  else if (sComputerRandomChoice === "scissors" && userInput === "scissors") {
+    userWin = 0.5;
+  } else if (sComputerRandomChoice === "paper" && userInput === "paper") {
+    userWin = 0.5;
+  } else if (sComputerRandomChoice === "stone" && userInput === "stone") {
+    userWin = 0.5;
+  }
```
Comment (lee-kimmixq) [inline project1-sps/script.js:58]: Could we try to shorten this by using || (OR) operators? Think something like grouping all those with the same userWin result and separating the conditions with || instead. e.g. lines 52-56 can be `else if ((com === "scissors" && user === "scissors") || (com === "paper" && user === "paper") || (com === "stone" && user === "stone"))`

File: project1-sps/script.js [hunk 9]
```diff
@@ -1,4 +1,144 @@
-var main = function (input) {
-  var myOutputValue = 'hello world';
-  return myOutputValue;
+// initialise variables
+let boolUserEnteredName = 0;
```
Comment (lee-kimmixq) [inline project1-sps/script.js:2]: Good application of a boolean function here. But could we make our code more efficient by omitting this variable? Maybe a test if the username string is empty i.e. `if (userName === "")` will serve the same purpose!

File: project1-sps/index.html [hunk 10]
```diff
@@ -72,7 +72,12 @@ <h1 id="header">🚀 Basics Project 1: Scissors Paper Stone 🚀</h1>
       <!-- The <button> tag represents.. you guessed it! a button. -->
       <button id="submit-button">Submit</button>
       <p>Output:</p>
-      <div class="output-div" id="output-div"></div>
+      <div class="output-div" id="output-div">
+        Welcome to play this Scissors Paper Stone game. <br />
+        <br />
+        Please enter your Name in the "Input" box and click "Submit".
+      </div>
```
Comment (lee-kimmixq) [inline project1-sps/index.html:79]: Nice :)

**Review by lee-kimmixq — COMMENTED** [review]
Hey Kaiquan! Great work on this project :) You've definitely understood all the concepts we've went through so far very well and managed to apply them in your code. Your code is clean and easy to read (with the help of your comments) in general. Naming follows the conventions and are appropriate too. I've provided some suggestions on how to refactor/make your code shorter - take these just as a reference (there's not clear right or wrong in coding)

Just some other things I want to highlight:
- You don't have to use the same name for variables in the input/output of your helper function and inside your main functions. Sometimes another name might be more appropriate.
- For input validations, maybe early returns will make your code shorter and reduce the number of boolean variables that you need to keep track of (like `userWrongInput`).

File: project1-sps/script.js (hunk 1 above)
Comment (KaiquanMah) [reply project1-sps/script.js:66]: if user has not entered name  (boolUserEnteredName === 0) but input is also empty (userInput = "")
- the code block underneath this IF statement will not be executed
- The output will show 'undefined'
- boolUserEnteredName will remain as 0


I used '&&' to workaround the issue of using '||' (OR) in the expression. 
- My intention is to block the user from entering the next phase/mode of the game until he/she has provided a name
- but this introduced the side effect where the output remains as 'undefined' until the user enters a username


If '||' is used instead and a user does not input anything and clicks on the 'submit' button, 
- boolUserEnteredName was 0, so "(boolUserEnteredName === 0 || userInput != "")"  would have evaluated to True
- the boolUserEnteredName value inside the code block would be updated to 1, while the user has not entered a name
- so the user will be able to proceed with the game without a game
- which was not my original intention

File: project1-sps/script.js (hunk 5 above)
Comment (KaiquanMah) [reply project1-sps/script.js:142]: Interesting tip Kimmi. This is helpful to learn.

I am visualising that in the makeResultMessage(userWin) helper function, we can slowly construct 
- some common lines and substitute in the corresponding variables
- then use IF-ELSE IF-ELSE statements to add on the scenario/expression-specific lines

File: project1-sps/script.js (hunk 8 above)
Comment (KaiquanMah) [reply project1-sps/script.js:58]: Thank you for sharing this tip Kimmi.

File: project1-sps/script.js (hunk 9 above)
Comment (KaiquanMah) [reply project1-sps/script.js:2]: Sure. Maybe if I added some variable to assign/update and keep track the "mode of the game" in the code, we can evaluate  the 'userInput' into the main function directly within the code blocks for each game mode?

**Comment by KaiquanMah** [comment]
2022.04.23 Commit includes:
Day 7 Inclass + Project 3 Blackjack (Base + Comfortable HTML/CSS items + visible/hidden buttons)
